### ISSUE TITLE
Cant get related resource

### ISSUE THREAD

**Issue by wayne-o** [scott-mcdonald/JsonApiFramework#22]
I'm sending the following PATCH from my UI:

```{"data":{"id":"4221fe2f-ca7b-425d-b45a-a09f4936a22b","attributes":{"contact-trade-name":"asdad","contact-name":"aasd","contact-phone":"asdad","contact-email":"asdad@adasd.asd","created-date":"2016-12-19T14:07:59.833","status":"AwaitingBillData"},"relationships":{"share-user":{"data":{"type":"users","id":"5078b2ee-2348-44f5-95a0-3c14c2ad2b1d"}}},"type":"jobs"}}```

And on my server i am doing the following:

```
using (var documentContext = new DocumentContext(HttpContext.Current.Request.Url, document))
        {
                    var job = documentContext.GetResource<Job>();

                    var relationships = documentContext.GetResourceRelationships(job);
                    var shareUserRel = relationships["share-user"];
                    var shareUser =  documentContext.GetRelatedResource<Data.ServiceModel.User>(shareUserRel);
        }
```

however `shareUser` is always `null` 

shareUserRel has the following when i inspect it:

```{ResourceIdentifier [type=users id=5078b2ee-2348-44f5-95a0-3c14c2ad2b1d]}```

if that helps.

Is there anything glaringly obvious that i might be doing wrong?

**Comment by scott-mcdonald** [comment]
Hi Wayne,

So I reformatted your client-side json:api document so it is easier to read:

````json
{
  "data": {
    "type": "jobs",
    "id": "4221fe2f-ca7b-425d-b45a-a09f4936a22b",
    "attributes": {
      "contact-trade-name": "asdad",
      "contact-name": "aasd",
      "contact-phone": "asdad",
      "contact-email": "asdad@adasd.asd",
      "created-date": "2016-12-19T14:07:59.833",
      "status": "AwaitingBillData"
    },
    "relationships": {
      "share-user": {
        "data": {
          "type": "users",
          "id": "5078b2ee-2348-44f5-95a0-3c14c2ad2b1d"
        }
      }
    }
  }
}
````

So to be clear your client application is sending this json:api document with the PATCH verb so you are requesting the server to do the following:
- Update an existing job by updating the following attributes:
    - contact-trade-name
    - contact-name
    - contact-phone
    - contact-email
    - created-date
    - status
- And set the "to-one-linkage" from the existing job with identifier `4221fe2f-ca7b-425d-b45a-a09f4936a22b` to an existing user with the identifier `5078b2ee-2348-44f5-95a0-3c14c2ad2b1d`.

So yes on the server-side when you call `GetRelatedResource` it should be `null` because there is no included resource in the `included` section of the json:api document. Nor should there be when creating a client-side document.

I think you are little confused on the "inclusion" of related resources so let me give you an example where `GetRelatedResource` method would return an actual resource:

Client issues a GET request to a hypermedia server, something like this:

    http://api.example.com/jobs/4221fe2f-ca7b-425d-b45a-a09f4936a22b?include=share-user

Hypermedia API server returns something like this:

````json
{
  "data": {
    "type": "jobs",
    "id": "4221fe2f-ca7b-425d-b45a-a09f4936a22b",
    "attributes": {
      "contact-trade-name": "asdad",
      "contact-name": "aasd",
      "contact-phone": "asdad",
      "contact-email": "asdad@adasd.asd",
      "created-date": "2016-12-19T14:07:59.833",
      "status": "AwaitingBillData"
    },
    "relationships": {
      "share-user": {
        "data": {
          "type": "users",
          "id": "5078b2ee-2348-44f5-95a0-3c14c2ad2b1d"
        }
      }
    }
  },
  "included": [
    {
      "type": "users",
      "id": "5078b2ee-2348-44f5-95a0-3c14c2ad2b1d",
      "attributes": {
        "first-name": "John",
        "last-name": "Doe",
        "twitter": "jdoe"
      }
    }
  ]
}
````

Now on the client-side you "read" the json:api document you will get a non-null "share-user" related resource.

````csharp
using (var documentContext = new DocumentContext(HttpContext.Current.Request.Url, document))
{
    var job = documentContext.GetResource<Job>();
    var relationships = documentContext.GetResourceRelationships(job);
    var shareUserRel = relationships["share-user"];
    var shareUser =  documentContext.GetRelatedResource<Data.ServiceModel.User>(shareUserRel);
    Assume(shareUser != null);
}
````
 
So in summary, your original client-side document did not have the included user related to the job resource. In fact client-side document do not include related resources when building documents, only server-side document building does either through a client include or server-side auto including of resources.

Hope that clarifies your issue...

**Comment by wayne-o** [comment]
That is excellent :) I did start giving the specs a proper read last night also - thank you for all of this - and your patience :) 

The issue I face now is that I am retrieving the initial data from a resource collection and as far as I can see there is no `includes` available here - i'm guessing for perf reasons? If i'm right I guess the best way to circumvent it would be to change my service model into a flatter structure so that the include isn't required for this end point - so maybe allow `shareUserId` on the collection GET.

**Comment by scott-mcdonald** [comment]
I guess I do not understand exactly what your issue is with the "includes" and related resources. If we are talking about client-side, the only time you build a document is to either create a new resource (POST) or update an existing resource (PATCH).

When you are creating new resources you can only create a json:api document with 1 resource as the primary data because of the json:api specification states you **MUST** only include a single resource as the primary data and basically create 1 resource per POST call.
http://jsonapi.org/format/#crud-creating

Likewise for updating existing resources with PATCH.
http://jsonapi.org/format/#crud-updating

Either POST or PATCH does support creating/updating resources with resource linkage to already existing resources. Anyways there is never a need to add "included" resources to a client-side document to POST or PATCH as the payload to a server.

If my explanation doesn't resolve your problem, please show me with an example so I can better understand your requirements...

**Comment by wayne-o** [comment]
OK - so I have a list of jobs on my UI and they come from calling /jobs which builds the JSON using ResourceCollection:

              ```
var jobServiceModel = jobs.ToList().Select(x => _jobFactory.CreateJob(x));
                    var document = documentContext
                        .NewDocument(HttpContext.Current.Request.Url)
                        .Links()
                        .AddUpLink()
                        .AddSelfLink()
                        .LinksEnd()
                        .ResourceCollection(jobServiceModel)
                            .Relationships()
                                .AddRelationship("owner-user", Keywords.Related, Keywords.Self)
                                .AddRelationship("share-user", Keywords.Related, Keywords.Self)
                            .RelationshipsEnd()
                            .Links()
                                .AddSelfLink()
                            .LinksEnd()
                        .ResourceCollectionEnd()
                        .WriteDocument();

                    return this.Ok(document);
```

Now - I was hoping to see if the job resource from inside the UI - has a shareUser - if it does - show a button 'Cancel Share' for that job - if not then show a button 'Share' for that job. 

I am using Ember for my UI and once it has the job resources in its model it thinks they are loaded so if i fetch the jobs from the collection endpoint as above - it will just think that they do not have any share users because I can't include the share user in the JSON because there is no Include avaliable for ResourceCollection.

Does that make sense?

**Comment by wayne-o** [comment]
The second part of the issue is that from the UI (Ember) I was hoping to say `job.set('shareUser', user)` and then save that - which is how I got to the point i am in on my 1st post :p

**Comment by wayne-o** [comment]
Would it be easier / ok if I shared the repo / site with you?

**Comment by scott-mcdonald** [comment]
For point 1, auto-include the share-user from the server-side so the client can look at the share-user relationship and query if the resource linkage is present or not. Something like the following:

````csharp
var job = ...
var jobShareUser = ...

var document = documentContext
    .NewDocument(HttpContext.Current.Request.Url)
        .Links()
            .AddUpLink()
            .AddSelfLink()
        .LinksEnd()
        .Resource(job)
            .Relationships()
                .AddRelationship("owner-user", Keywords.Related, Keywords.Self)
                .AddRelationship("share-user", Keywords.Related, Keywords.Self)
            .RelationshipsEnd()
            .Links()
                .AddSelfLink()
            .LinksEnd()
        .ResourceEnd()
        .Included()
            .AddToOne(ToOneResourceLinkage.Create(job, "share-user", jobShareUser))
        .IncludedEnd()
    .WriteDocument();

return this.Ok(document);
````

For point 2, if you are building a client-side PATCH document with JsonApiFramework to either set or clear the related share-user try the following:

Clear any current existing user as the related "share-user" from a job

````csharp
var job = ...
var jobShareUser = ...

var document = documentContext
    .NewDocument()
        .Resource<Job>()
            .SetId(job.Id)
            .Relationships()
                .AddRelationship("share-user")
            .RelationshipsEnd()
         .ResourceEnd()
    .WriteDocument();
````

Set an existing user as the related "share-user" to a job

````csharp
var job = ...
var jobShareUser = ...

var document = documentContext
    .NewDocument()
        .Resource<Job>()
            .SetId(job.Id)
            .Relationships()
                .AddRelationship("share-user", jobShareUser.Id)
            .RelationshipsEnd()
         .ResourceEnd()
    .WriteDocument();
````

Note on the server side the related resource identifier will either be an actual resource identifier (set) or null (clear).

**Comment by wayne-o** [comment]
Hi mate - what would the JSON being posted to the server look like when you send the result of the following:

```var job = ...
var jobShareUser = ...

var document = documentContext
    .NewDocument()
        .Resource<Job>()
            .SetId(job.Id)
            .Relationships()
                .AddRelationship("share-user", jobShareUser.Id)
            .RelationshipsEnd()
         .ResourceEnd()
    .WriteDocument();```

Because that's basically what i have for my initial post - just in ember-data form.

**Comment by wayne-o** [comment]
Ideall I would like to have the `included` property on this lot that comes from the ResourceCollection

```
{
  "links": {
    "up": "https://localhost:44314/api",
    "self": "https://localhost:44314/api/jobs"
  },
  "data": [
    {
      "type": "jobs",
      "id": "cec9da3d-7a87-46e8-9a86-0941d89e426a",
      "attributes": {
        "contact-name": "asdasda",
        "contact-trade-name": "asdasd",
        "contact-phone": "asdasd",
        "contact-email": "asdasd@asd.asd",
        "is-accepting-card-payments": false,
        "is-pos": false,
        "created-date": "2016-12-20T08:11:47.773",
        "last-modified-date": "2016-12-20T09:36:39.033",
        "is-deleted": false,
        "status": "AwaitingBillData",
        "owner-user": null,
        "share-user": null,
        "is-shared": true
      },
      "relationships": {
        "owner-user": {
          "links": {
            "related": "https://localhost:44314/api/jobs/cec9da3d-7a87-46e8-9a86-0941d89e426a/owner-user",
            "self": "https://localhost:44314/api/jobs/cec9da3d-7a87-46e8-9a86-0941d89e426a/relationships/owner-user"
          }
        },
        "share-user": {
          "links": {
            "related": "https://localhost:44314/api/jobs/cec9da3d-7a87-46e8-9a86-0941d89e426a/share-user",
            "self": "https://localhost:44314/api/jobs/cec9da3d-7a87-46e8-9a86-0941d89e426a/relationships/share-user"
          }
        }
      },
      "links": {
        "self": "https://localhost:44314/api/jobs/cec9da3d-7a87-46e8-9a86-0941d89e426a"
      }
    },
    {
      "type": "jobs",
      "id": "8d922678-bd4c-4199-8260-ef063b4b070a",
      "attributes": {
        "contact-name": "ppp",
        "contact-trade-name": "ppp",
        "contact-phone": "ppp",
        "contact-email": "ppp@asdasd.asd",
        "is-accepting-card-payments": false,
        "is-pos": false,
        "created-date": "2016-12-20T08:11:28.94",
        "last-modified-date": "2016-12-20T09:36:44.77",
        "is-deleted": false,
        "status": "AwaitingBillData",
        "owner-user": null,
        "share-user": null,
        "is-shared": true
      },
      "relationships": {
        "owner-user": {
          "links": {
            "related": "https://localhost:44314/api/jobs/8d922678-bd4c-4199-8260-ef063b4b070a/owner-user",
            "self": "https://localhost:44314/api/jobs/8d922678-bd4c-4199-8260-ef063b4b070a/relationships/owner-user"
          }
        },
        "share-user": {
          "links": {
            "related": "https://localhost:44314/api/jobs/8d922678-bd4c-4199-8260-ef063b4b070a/share-user",
            "self": "https://localhost:44314/api/jobs/8d922678-bd4c-4199-8260-ef063b4b070a/relationships/share-user"
          }
        }
      },
      "links": {
        "self": "https://localhost:44314/api/jobs/8d922678-bd4c-4199-8260-ef063b4b070a"
      }
    }
  ]
}
```

**Comment by wayne-o** [comment]
Or to be clear - there is no way to do this:

```
var jobServiceModel = jobs.ToList().Select(x => _jobFactory.CreateJob(x));
                    var document = documentContext
                        .NewDocument(HttpContext.Current.Request.Url)
                        .Links()
                        .AddUpLink()
                        .AddSelfLink()
                        .LinksEnd()
                        .ResourceCollection(jobServiceModel)
                            .Relationships()
                                .AddRelationship("owner-user", Keywords.Related, Keywords.Self)
                                .AddRelationship("share-user", Keywords.Related, Keywords.Self)
                            .RelationshipsEnd()
                            .Links()
                                .AddSelfLink()
                            .LinksEnd()
                        .ResourceCollectionEnd()
                        .Included()
                            .ToOne(jobServiceModel, "share-user", jobs.ToList().Select(x=> _userFactory.CreateUser(x.ShareUser)))
                        .IncludedEnd()
                        .WriteDocument();
```

**Comment by wayne-o** [comment]
OK so i have seen that i can build up a `List<IToOneResourceLinkage<Job, User>>` to be passed to `ToOne` like so:
```
var jobServiceModel = jobs.ToList().Select(x => _jobFactory.CreateJob(x));

                    var links = new List<IToOneResourceLinkage<Job, User>>();

                    foreach (var job in jobServiceModel)
                    {
                        if (job.ShareUserId != null && job.ShareUserId != Guid.NewGuid())
                        {
                            {
                                var shareUser =
                                    jobs.Select(x => x.ShareUser).FirstOrDefault(x => x.Id == job.ShareUserId);
                                links.Add(ToOneResourceLinkage.Create(job, "share-user",
                                    _userFactory.CreateUser(shareUser)));
                            }
                        }
                    }

                    var document = documentContext
                        .NewDocument(HttpContext.Current.Request.Url)
                        .Links()
                        .AddUpLink()
                        .AddSelfLink()
                        .LinksEnd()
                        .ResourceCollection(jobServiceModel)
                            .Relationships()
                                .AddRelationship("owner-user", Keywords.Related, Keywords.Self)
                                .AddRelationship("share-user", Keywords.Related, Keywords.Self)
                            .RelationshipsEnd()
                            .Links()
                                .AddSelfLink()
                            .LinksEnd()
                        .ResourceCollectionEnd()
                        .Included()
                            .ToOne(links)
                            .ToOneEnd()
                        .IncludedEnd()
                        .WriteDocument();
```

The issue being that if more than one job is shared to a single user I get the following exception:

```
{
  "Message": "An error has occurred.",
  "ExceptionMessage": "An item with the same key has already been added.",
  "ExceptionType": "System.ArgumentException",
  "StackTrace": "..."
}
```

**Comment by wayne-o** [comment]
To get this to work I had to amend `DocumentBuilderContext` like so:

```
public void AddDomReadWriteResource(ResourceIdentifier domResourceKey, DomReadWriteResource domReadWriteResource)
        {
            Contract.Requires(domResourceKey != null);
            Contract.Requires(domReadWriteResource != null);
            if (ContainsDomReadWriteResource(domResourceKey))
            {
                return;
            }

            this.DomReadWriteResourceDictionary.Add(domResourceKey, domReadWriteResource);
        }
```

I'm wondering if this is just patching up my issue though

**Comment by scott-mcdonald** [comment]
I think this is what you want:

From your example you have something like the following:

Jobs and Users coming from the "database" which I'll give the type names JobDbEntity and UserDbEntity with JobDbEntity having a navigation property called ShareUser, etc.

You also have "Factory" classes to create JobResource and UserResource types which are part of your hypermedia API service model schema.

````csharp
var jobEntities = Get collection of JobDbEntity with related UserDbEntity if there is one.
var jobFactory = Create JobResource from JobDbEntity
var userFactory = Create UserResource from UserDbEntity

var jobAndRelatedShareUserCollection = jobEntities
  .Select(x =>
  {
    var jobResource = jobFactory.Create(x);
    var jobToUserResourceLinkage = x.ShareUser != null
      ? ToOneResourceLinkage.Create(jobResource, "share-user", userFactory.Create(x.ShareUser)
      : null;
    return new Tuple<JobResource, IToOneResourceLinkage<JobResource, UserResource>>(jobResource, jobToUserResourceLinkage);
  }
  .ToList();

var jobResourceCollection = jobAndRelatedShareUserCollection
  .Select(x => x.Item1)
  .ToList();

var jobToUserResourceLinkageCollection = jobAndRelatedShareUserCollection
  .Where(x => x.Item2 != null)
  .Select(x => x.Item2)
  .ToList();

var document = documentContext
    .NewDocument(HttpContext.Current.Request.Url)
        .Links()
            .AddUpLink()
            .AddSelfLink()
        .LinksEnd()
        .ResourceCollection(jobResourceCollection)
            .Relationships()
                .AddRelationship("share-user", Keywords.Related, Keywords.Self)
            .RelationshipsEnd()
            .Links()
                .AddSelfLink()
            .LinksEnd()
        .ResourceCollectionEnd()
        .Included()
            .AddToOne(jobToUserResourceLinkageCollection)
        .IncludedEnd()
    .WriteDocument();

return this.Ok(document);
````

The framework should handle correctly if the same user is included by multiple job resources. If not, then that is a defect in the framework.

**Comment by krilbert** [comment]
Sorry to interrupt but I have a similar question and hope you guys can help me 😳 . I have a list of clients and each one has a list of contacts. The question would be, how can I get the `included` part? I mean, get the relationships (data) on each client and the full contact resource on the included section.

```c#
var list = new Clients().GetAll(userId);

var currentRequestUrl = HttpContext.Current.Request.Url;
using (var documentContext = new RCServiceDocumentContext(currentRequestUrl))
{
  var meta = Meta.Create(new
  {
    count = list.Count,
    offset = 0,
    limit = list.Count,
    totalCount = list.Count
  });

  var document = documentContext
      .NewDocument(currentRequestUrl)
          .SetJsonApiVersion(JsonApiVersion.Version10)
          .SetMeta(meta)
          .ResourceCollection(list)
          .ResourceCollectionEnd()
          .Included()
          .IncludedEnd()
          .WriteDocument();

  return Ok(document);
}
```
Also I include the way I understood to add the `meta` but it doesn't recognize the `JsonFormatter.SerializerSettings`.

```c#
  public class ClientConfiguration : ResourceTypeBuilder<Client>
  {
    public ClientConfiguration()
    {
      // Attributes
      this.Attribute(x => x.Contacts).Ignore();

      // Relationships
      ToManyRelationship<Contact>("contacts").SetApiRelPathSegment("contacts");
      ToOneRelationship<Relationship>("relationship");
      ToOneRelationship<Industry>("industry");
    }
  }

  public class ContactConfiguration : ResourceTypeBuilder<Contact>
  {
  }
```

Thanks in advance for any comment 😄 
> [Source Code](https://github.com/krilbert/macWebAPI/blob/master/rcAPI/Controllers/ClientsController.cs)

**Comment by scott-mcdonald** [comment]
Looking at your source code, you are going to want something like the following:

```csharp

var clientCollection = new Clients().GetAll(userId);

var clientToIndustryCollection = clientCollection
    .Select(client => ToOneResourceLinkage.Create(client, "industry", client.Industry)
    .ToList();

var clientToRelationshipCollection = clientCollection
    .Select(client => ToOneResourceLinkage.Create(client, "relationship", client.Relationship)
    .ToList();

var clientToContactsCollection = clientCollection
    .Select(client => ToManyResourceLinkage.Create(client, "contacts", client.Contacts)
    .ToList();

var document = documentContext
    .NewDocument(HttpContext.Current.Request.Url)
        .Links()
            .AddUpLink()
            .AddSelfLink()
        .LinksEnd()
        .ResourceCollection(clientCollection)
            .Relationships()
                .AddRelationship("industry", Keywords.Related, Keywords.Self)
                .AddRelationship("relationship", Keywords.Related, Keywords.Self)
                .AddRelationship("contacts", Keywords.Related, Keywords.Self)
            .RelationshipsEnd()
            .Links()
                .AddSelfLink()
            .LinksEnd()
        .ResourceCollectionEnd()
        .Included()

            .ToOne(clientToIndustryCollection)
                .Links()
                    .AddSelfLink()
                .LinksEnd()
            .ToOneEnd()

            .ToOne(clientToRelationshipCollection)
                .Links()
                    .AddSelfLink()
                .LinksEnd()
            .ToOneEnd()

            .ToMany(clientToContactsCollection)
                .Links()
                    .AddSelfLink()
                .LinksEnd()
            .ToManyEnd()

        .IncludedEnd()
    .WriteDocument();

return this.Ok(document);

Here are some notes from my quick look at your source code:

You never want to expose foreign keys so make sure you "ignore" foreign keys when designing your resource types. Also resources can not be composed of other resources, you expose other related resources through relationships not as aggregate attributes. Therefore an example Client configuration:

public class ClientConfiguration : ResourceTypeBuilder<Client>
{
    public ClientConfiguration()
    {
        // Ignore foreign keys
        this.Attribute(x => x.Id).Ignore();
        this.Attribute(x => x.User_id).Ignore();
        this.Attribute(x => x.Custom_client_id).Ignore();
        
        // Ignore aggregate resources
        this.Attribute(x => x.Industry).Ignore();
        this.Attribute(x => x.Relationship).Ignore();
        this.Attribute(x => x.Contacts).Ignore();
        
        // Relationships
        this.ToOneRelationship<Industry>("industry");
        this.ToOneRelationship<Relationship>("relationship");
        this.ToManyRelationship<Contact>("contacts");
    }
}
````

**Comment by krilbert** [comment]
Thank you very much for your time, I will try right now ^^
I just get this:
<img width="653" alt="screen shot 2016-12-22 at 8 37 02 pm" src="https://cloud.githubusercontent.com/assets/2858018/21446011/a8156348-c886-11e6-9cf8-3f50f105c30d.png">

**Comment by krilbert** [comment]
Sorry to bother but, how do you recommend me to extract the info from the database? do you know a better way? do I have to do the pagination by myself?

**Comment by krilbert** [comment]
Working like a charm! 😄

Just wondering, what I mentioned about the Meta serializer settings, is that the normal behavior?

**Comment by wayne-o** [comment]
This framework is awesome btw - once you get up and running with it - it's immense :)

**Comment by scott-mcdonald** [comment]
@krilbert Can you explain with more detail what the Meta serializer settings issue actually is?

**Comment by krilbert** [comment]
@scott-mcdonald sorry for the delay, I was out of town for the holidays.

What I was talking about is that in my `WebApiConfig.cs` I specify the serializer settings, like this:
```c#
config.Formatters.JsonFormatter.SerializerSettings = Shared.SerializerSettings;

public static JsonSerializerSettings SerializerSettings => new JsonSerializerSettings
{
  Formatting = Formatting.None,
  NullValueHandling = NullValueHandling.Ignore,
  ContractResolver = new CamelCasePropertyNamesContractResolver()
};
```
And in my controller I need to add a `meta` property so I did this:
```c#
        var meta = Meta.Create(new
        {
          count = clientCollection.Count,
          offset = 0,
          limit = clientCollection.Count,
          totalCount = clientCollection.Count
        });

        document = documentContext
          .NewDocument(currentRequestUrl)
            .SetJsonApiVersion(JsonApiVersion.Version10)
            .SetMeta(meta)
          .WriteDocument();
```
But looks like my `JsonFormatter.SerializerSettings` are being ignored. I mean is not converting to CamelCase and including null values. Should I specify them again somewhere else?

**Comment by krilbert** [comment]
Sorry to bother again, now that I have more than one record, I'm getting `An item with the same key has already been added.` I tried with Distinct but nothing 😳

**Comment by scott-mcdonald** [comment]
@krilbert I was away on vacation as well and just got back yesterday. I will take a look at the serializer settings being ignored when creating Meta objects. Also can you provide an example or link to the source code that demonstrates the error `An item with the same key has already been added`...

**Comment by krilbert** [comment]
Thks a lot for your quick response. Sure thing, I believe you could see it with this repo: `https://github.com/krilbert/macWebAPI` thks in advance

**Comment by krilbert** [comment]
```c#
        var clientToIndustryCollection = clientCollection
          .Select(client => ToOneResourceLinkage.Create(client, "industry", client.Industry))
          .ToList();

        document = documentContext
          .NewDocument(currentRequestUrl)
            .SetJsonApiVersion(JsonApiVersion.Version10)
            .ResourceCollection(clientCollection)
              .Relationships()
                .Relationship("industry").RelationshipEnd()
              .RelationshipsEnd()
            .ResourceCollectionEnd()
            .Included()
              .ToOne(clientToIndustryCollection).ToOneEnd()  // The exception is being thrown here
            .IncludedEnd()
          .WriteDocument();
```
`clientToIndustryCollection` has repetitive value.
Sorry for my english, is not my native language 😳

**Comment by krilbert** [comment]
I believe the problem occurs on `DocumentBuilder.cs` line `286` When
```c#
var clientToIndustryCollection = clients
     .Select(client => ToOneResourceLinkage.Create(client, "industry", client.Industry))
     .ToList();
```
has two equal items (resources) from different client. I think the way I did it produce that, I'll try another approach

**Comment by scott-mcdonald** [comment]
@krilbert I verified that indeed the "clientToIndustryCollection" issue your brought up is a bug in JsonApiFramework. Conceptually there is nothing wrong with 2 clients having the same industry - perfectly valid and I can reproduce the exception.

I am going to write up a separate bug ticket, write a unit test, then fix in the source code. Thanks for the catch...

**Comment by scott-mcdonald** [comment]
@krilbert See #24 for an enhancement to your Meta creation issue not honoring the current JsonSerializerSettings you defined for the WebApi project. 

Fixes for #23 and #24 will be in the next release of JsonApiFramework v1.1.3-beta.

Also going to close this issue as the original defect by @wayne-o is resolved. Please create new issues as they are discovered.

**Comment by krilbert** [comment]
Already test it :D thks, I'm glad I can help

**Comment by scott-mcdonald** [comment]
@krilbert Make sure you download the latest version of JsonApiFramework 1.1.3-beta. Create Meta objects with your shared serializer settings as well like I documented in a previous comment. I did clone your repository and verified JsonApiFramework 1.1.3-beta resolved your two issues you found. #23 was the big catch, thanks again! Good luck!
